Std/Data/BitVec/Basic.lean

```diff
@@ -0,0 +1,185 @@
+instance : Div (BitVec n) := ⟨BitVec.div⟩
+instance : LT (BitVec n)  := ⟨fun x y => BitVec.lt x y⟩
+instance : LE (BitVec n)  := ⟨fun x y => BitVec.le x y⟩
+
+/-- Bitwise and for bit vectors. -/
+protected def and (x y : BitVec n) : BitVec n :=
+  { val := x.val &&& y.val }
+/-- Bitwise or for bit vectors. -/
+protected def or (x y : BitVec n) : BitVec n :=
+  { val := x.val ||| y.val }
+/-- Bitwise xor for bit vectors. -/
+protected def xor (x y : BitVec n) : BitVec n :=
+  { val := x.val ^^^ y.val }
+/-- Complement for bit vectors. -/
+protected def complement (x : BitVec n) : BitVec n :=
+  0 - (x + .ofNat n 1)
+/-- Shift left for bit vectors. -/
+protected def shiftLeft (a : BitVec n) (s : Nat) : BitVec n :=
+  .ofNat n (a.toNat <<< s)
+/-- Shift right for bit vectors. -/
+protected def shiftRight (a : BitVec n) (s : Nat) : BitVec n :=
+  .ofNat n (a.toNat >>> s)
+
+instance : Complement (BitVec w) := ⟨BitVec.complement⟩
+instance : AndOp (BitVec w) := ⟨BitVec.and⟩
+instance : OrOp (BitVec w) := ⟨BitVec.or⟩
+instance : Xor (BitVec w) := ⟨BitVec.xor⟩
+instance : HShiftLeft (BitVec w) Nat (BitVec w) := ⟨BitVec.shiftLeft⟩
+instance : HShiftRight (BitVec w) Nat (BitVec w) := ⟨BitVec.shiftRight⟩
+
+/-- Rotate left. -/
+def rotateLeft (x : BitVec w) (n : Nat) : BitVec w :=
+  x <<< n ||| x >>> (w - n)
+
+/-- Rotate right. -/
+def rotateRight (x : BitVec w) (n : Nat) : BitVec w :=
+  x >>> n ||| x <<< (w - n)
+
+/-- Append. -/
+def append (a : BitVec n) (b : BitVec m) : BitVec (n+m) :=
+  .ofNat (n+m) (a.toNat <<< m + b.toNat)
+
+instance : HAppend (BitVec w) (BitVec v) (BitVec (w+v)) := ⟨BitVec.append⟩
+/--
+Extraction of bits `i` down to `j` from a bit vector of size `n` to yield a
+new bitvector of size `i - j + 1`
+-/
+def extract (i j : Nat) (a : BitVec n) : BitVec (i - j + 1) :=
+  BitVec.ofNat _ (a.toNat >>> j)
+
+/--
+`repeat_ i x` means concatenate `i` copies of `x`.
+Recall that `repeat` is a keyword in Lean.
+-/
+def repeat_ : (i : Nat) → BitVec w → BitVec (w*i)
+  | 0,   _ => 0
+  | n+1, x =>
+    have hEq : w + w*n = w*(n + 1) := by
+      rw [Nat.mul_add, Nat.add_comm, Nat.mul_one]
+    hEq ▸ (x ++ repeat_ n x)
+
+/-- Zero extension. -/
+def zeroExtend (i : Nat) (x : BitVec w) : BitVec (w+i) :=
+  have hEq : w+i = i+w := Nat.add_comm _ _
+  hEq ▸ ((0 : BitVec i) ++ x)
+
+/-- Signed extension. -/
+def signExtend (i : Nat) (x : BitVec w) : BitVec (w+i) :=
+  have hEq : ((w-1) - (w-1) + 1)*i + w = w+i := by
+    rw [Nat.sub_self, Nat.zero_add, Nat.one_mul, Nat.add_comm]
+  hEq ▸ ((repeat_ i (extract (w-1) (w-1) x)) ++ x)
+
+/-- Return a prefix of size `v` of bit vector `x`. -/
+def shrink (v : Nat) (x : BitVec w) : BitVec v :=
+  if hZero : 0 < v then
+    have hEq : v - 1 + 0 + 1 = v := by
+      rw [Nat.add_zero]
+      exact Nat.sub_add_cancel hZero
+    hEq ▸ x.extract (v - 1) 0
+  else 0
+
+/-- Return the `i`-th least significant bit. -/
+def lsbGet (x : BitVec w) (i : Nat) : Bool :=
+  x.extract i i != 0
+
+end Std.BitVec
```

Comment:
```suggestion
```

for better or worse, the current Std convention is not close namespaces at the end of the file.